Simplify the K8sExecutor and K8sPodOperator

dimberman opened the pull request:
As discussed in an earlier email thread, this PR removes much of the configuration elements of the K8sExecutor for a much simpler design.

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

turbaszek commented on airflow/kubernetes/pod_generator.py:
```diff
@@ -96,100 +96,18 @@ class PodGenerator:
         pod_template_file: Optional[str] = None,
-        extract_xcom: bool = False,
-        priority_class_name: Optional[str] = None,
+        extract_xcom=True
```
```suggestion
        extract_xcom: bool =True
```

turbaszek commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -175,7 +175,7 @@ def __init__(self,  # pylint: disable=too-many-arguments,too-many-locals
                  config_file: Optional[str] = None,
                  node_selectors: Optional[Dict] = None,
-                 image_pull_secrets: Optional[str] = None,
+                 image_pull_secrets: List[k8s.V1LocalObjectReference] = None,
```
```suggestion
                 image_pull_secrets: Optional[List[k8s.V1LocalObjectReference]] = None,
```

turbaszek commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -365,44 +365,57 @@ def create_new_pod_for_operator(self, labels, launcher) -> Tuple[State, k8s.V1Po
-            pod=self.full_pod_spec,
-        ).gen_pod()
+        from kubernetes.client import models as k8s
+        pod = None
```
```suggestion

```
Probably no need for that as we have if/else

turbaszek commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -365,44 +365,57 @@ def create_new_pod_for_operator(self, labels, launcher) -> Tuple[State, k8s.V1Po
-            pod_template_file=self.pod_template_file,
-            pod=self.full_pod_spec,
-        ).gen_pod()
+        from kubernetes.client import models as k8s
```
Any particular reason to import this locally?

dimberman replied:
Nope removed that

turbaszek reviewed (commented):
-2.5k of lines - love it ❤️ This is still WIP but we should remember about note in UPDATING.md

dimberman commented on tests/kubernetes/test_pod_generator.py:
```diff
@@ -226,238 +186,169 @@ def test_gen_pod_extract_xcom(self, mock_uuid):
-                }],
-            }
-        }, result)
+    def test_from_obj(self):
```
@kaxil @ashb @potiuk @turbaszek one remaining question is what do we do about the executor_config? Do we force users to inject a k8s.V1Pod object for their executor config and just manually delete a few fields (like pod names)? Or do we continue to allow this dictionary.

kaxil commented on airflow/example_dags/example_kubernetes_executor_config.py:
```diff
@@ -52,60 +53,67 @@ def test_volume_mount():
         task_id="start_task",
         python_callable=print_stuff,
-        executor_config={
-            "KubernetesExecutor": {
```
I think having the executor name would be helpful if in future we want to see configs for multiple executors

dimberman replied:
The problem I'm seeing is that k8s.V1Pod objects aren't serializable into dictionaries. Not sure how to get around that and would rather not optimize for a future I don't think will come

dimberman replied:
(at this point the executor_config has existed for years and no other executor has used it. If it wasn't such a pain I'd push for us to change the name to something like "kubernetes_modifier_pod")

kaxil replied:
> The problem I'm seeing is that k8s.V1Pod objects aren't serializable into dictionaries. Not sure how to get around that and would rather not optimize for a future I don't think will come

We will need it for DAG Serialization too, is it not possible to convert V1Pod to dict ? I thought each k8s object had some kind of method for that

dimberman replied:
Let me try again. Maybe I'm missing something.

dimberman replied:
```  
File "/home/airflow/.local/lib/python3.7/site-packages/airflow/utils/json.py", line 54, in _default
    raise TypeError(f"Object of type '{obj.__class__.__name__}' is not JSON serializable")
TypeError: Object of type 'V1Pod' is not JSON serializable
```

Yeah so if I try to make the pod a field I end up getting this error when I try to render on the page. Not sure what the best way forward is there. Maybe add a field? I don't think we need to change the DB as the executor_config isn't stored in the DB anyways.

dimberman replied:
(though the downside of adding a field is that it would create even more confusing in transitioning DAGs. I'd personally either figure out a solution to this json serialization issue, or just have multiple types for the executor_config)

kaxil replied:
We do store executor_config in TaskInstance table:

https://github.com/apache/airflow/blob/d7602654526fdd2876466371404784bd17cfe0d2/airflow/models/taskinstance.py#L222

Although we use `dill` to store it

kaxil replied:
`k8s.V1Pod(...).to_dict` should work for serialization, isn't it?

https://github.com/kubernetes-client/python/blob/v11.0.0/kubernetes/client/models/v1_pod.py#L179-L201

dimberman replied:
@kaxil yeah but forcing using to use the to_dict function seems kinda clunky, no?

We store the executor_config in the DB but TBH we don't really need to. Maybe we can just create new fields "kubernetes_executor_config" and "dag_serialization_config" and just remove the "executor_config" as a whole?

kaxil replied:
No, I meant, adding code in serialization module, https://github.com/apache/airflow/blob/e565368f2e988a06b0398b77e66859c138905ae7/airflow/serialization/serialized_objects.py#L159-L216

that if the object is k8s.V1Pod, store it as `k8s.V1Pod.to_dict()` .

kaxil replied:
>We store the executor_config in the DB but TBH we don't really need to. Maybe we can just create new fields "kubernetes_executor_config" and "dag_serialization_config" and just remove the "executor_config" as a whole?

If we don't need to store it in DB then sure we can remove it

dimberman replied:
Ok I got it working with the traditional executor_config. Thanks @kaxil!

kaxil commented on airflow/executors/kubernetes_executor.py:
```diff
@@ -70,116 +69,25 @@ class KubeConfig:  # pylint: disable=too-many-instance-attributes
+        self.worker_container_tag = conf.get(
+            self.kubernetes_section, 'worker_container_tag')
+        self.kube_image = '{}:{}'.format(
+            self.worker_container_repository, self.worker_container_tag)
```
```suggestion
        self.kube_image = f'{self.worker_container_repository}:{self.worker_container_tag}'
```

kaxil reviewed (commented):
Can you rebase on latest Master please if you haven't already?

kaxil commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -365,43 +351,53 @@ def create_new_pod_for_operator(self, labels, launcher) -> Tuple[State, k8s.V1Po
+        for secret in self.secrets:
+            pod = secret.attach_to_pod(pod)
+        if self.do_xcom_push:
+            from airflow.kubernetes.pod_generator import PodGenerator
+            pod = PodGenerator.add_xcom_sidecar(pod)
```
```suggestion
            pod = pod_generator.PodGenerator.add_xcom_sidecar(pod)
```

kaxil commented on telepresence.log:
```diff
@@ -0,0 +1,36 @@
+   0.0 TEL | Telepresence 0.105 launched at Mon Aug 24 20:51:33 2020
```
Don't think you wanted to add this file :)

dimberman committed:
Simplify Airflow on Kubernetes Story

Removes thousands of lines of code that essentially ammount to us
re-creating the Kubernetes API. Will offer a faster, simpler
KubernetesExecutor for 2.0

dimberman commented on airflow/kubernetes/pod_generator_deprecated.py:
```diff
@@ -0,0 +1,620 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
@ashb @potiuk @kaxil @turbaszek With this class we can actually maintain the existing executor_config with a deprecation warning until 2.1 or 2.2. This should help ease the transition. WDYT?

mik-laj commented on UPDATING.md:
```diff
@@ -153,6 +152,175 @@ The Old and New provider configuration keys that have changed are as follows
+For Airflow 2.0, the traditional `executor_config` will continue operation with a deprecation warning,
+but will be removed in a future version.
+
+### Changes to the KubernetesPodOperator
```
In my opinion, this tutorial will not allow easy migration between the old and the new representation. Many parameters are not described and the new representation will not always be obvious.  The user will have to read the version 1.10 code to understand what they should do now.

dimberman replied:
@mik-laj Thank you for that feedback. I've added more detailed docs on the changed parameters. Can you please take a look when you have some time?

mik-laj commented on UPDATING.md:
````diff
@@ -153,6 +152,175 @@ The Old and New provider configuration keys that have changed are as follows
+```
+We decided to keep the Secret class as users seem to really like that simplifies the complexity of mounting
+Kubernetes secrets into workers.
+
````
Can you add a migration guide for each configuration option?

dimberman replied:
Added!

mik-laj commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -65,11 +60,11 @@ class KubernetesPodOperator(BaseOperator):  # pylint: disable=too-many-instance-
     :type arguments: list[str]
     :param ports: ports for launched pod.
-    :type ports: list[airflow.kubernetes.pod.Port]
+    :type ports: list[k8s.V1ContainerPort]
```
I would be happy if there was an entry with an example for each parameter that has changed. We cannot assume that the user will know these two classes and will be able to migrate on their own.

ashb commented on airflow/utils/json.py:
```diff
@@ -20,6 +20,7 @@
 from datetime import date, datetime
 
 import numpy as np
+from kubernetes.client import models as k8s
```
```suggestion
try:
    from kubernetes.client import models as k8s
except ImportError:
    k8s = None
```

ashb commented on airflow/utils/json.py:
```diff
@@ -50,5 +51,8 @@ def _default(obj):
         elif isinstance(obj, (np.float_, np.float16, np.float32, np.float64,
                               np.complex_, np.complex64, np.complex128)):
             return float(obj)
+        elif isinstance(obj, k8s.V1Pod):
```
```suggestion
        elif k8s is not None and isinstance(obj, k8s.V1Pod):
```

ashb commented on airflow/serialization/enums.py:
```diff
@@ -42,3 +42,4 @@ class DagAttributeTypes(str, Enum):
     DICT = 'dict'
     SET = 'set'
     TUPLE = 'tuple'
+    POD = 'pod'
```
```suggestion
    POD = 'k8s.V1Pod'
```

Maybe?

mik-laj commented on tests/kubernetes/test_pod_generator.py:
```diff
@@ -241,238 +213,224 @@ def test_gen_pod_extract_xcom(self, mock_uuid):
                 }],
             }
         }, result)
+        # TODO: Should we save this feature?
```
?

mik-laj commented on tests/kubernetes/test_pod_generator.py:
```diff
@@ -241,238 +213,224 @@ def test_gen_pod_extract_xcom(self, mock_uuid):
-        result = PodGenerator.reconcile_pods(base_pod, mutator_pod)
-        self.assertEqual(base_pod, result)
+        pass
+        # mock_uuid.return_value = self.static_uuid
```
?

kaxil commented on airflow/example_dags/example_kubernetes_executor_config.py:
```diff
@@ -52,59 +53,68 @@ def test_volume_mount():
-            "KubernetesExecutor": {
-                "annotations": {"test": "annotation"}
-            }
+        executor_config={"podOverride": k8s.V1Pod(
```
After https://github.com/apache/airflow/pull/10756 is merged, we need to change `podOverride` -> `pod_override` in this PR

kaxil commented on airflow/example_dags/example_kubernetes_executor_config.py:
```diff
@@ -52,59 +53,68 @@ def test_volume_mount():
-                }
-            }
-        }
+        # executor_config={"podOverride": k8s.V1Pod(
```
is the intention to remove these comments?

kaxil commented on airflow/executors/kubernetes_executor.py:
```diff
@@ -483,17 +326,23 @@ def run_next(self, next_job: KubernetesJobType) -> None:
-            kube_executor_config=kube_executor_config,
-            worker_config=self.worker_configuration_pod
+            kube_mutation_object=kube_executor_config,
+            base_worker_pod=base_worker_pod
```
wherever this signature was changed, if we haven't already have it in Updating.md we should add it

dimberman replied:
@kaxil do we need to note this if it's not an external class?

ashb commented on kubernetes_tests/test_kubernetes_executor.yaml:
```diff
@@ -0,0 +1,82 @@
+  restartPolicy: Never
+  securityContext:
+    fsGroup: 50000
+    runAsUser: 50000
```
```suggestion
    runAsUser: '{{ .Values.uid }}'
    fsGroup: '{{ .Values.gid }}' 
```

etc.

ashb commented on chart/values.yaml:
```diff
@@ -541,32 +543,11 @@ config:
-    git_sync_run_as_user: '{{ .Values.uid }}'
-    git_ssh_known_hosts_configmap_name: '{{- if .Values.dags.gitSync.knownHosts }}{{ include "airflow_config" . }}{{ end }}'
-    git_ssh_key_secret_name: '{{- if .Values.dags.gitSync.sshKeySecret }}{{ .Values.dags.gitSync.sshKeySecret }}{{ end }}'
-    git_sync_credentials_secret: '{{- if .Values.dags.gitSync.credentialsSecret }}{{ .Values.dags.gitSync.credentialsSecret }}{{ end }}'
```
Didn't we discuss that the chart would take these helm values and generate the kube template file based on these?

dimberman replied:
@ashb OOF good catch thank you

dimberman replied:
@ashb every value in this section is brought from other parts of the helm file. Should we keep this around or delete this section and still have a default pod_template file from those other values (seeing as this section is specifically made for the airflow.cfg)?

ashb replied:
This section can go, yeah -- just using this as a place to comment on.

dimberman committed:
Fix podgen tests

davlum commented on airflow/kubernetes/pod_generator.py:
```diff
@@ -123,100 +124,18 @@ class PodGenerator:
-        priority_class_name: Optional[str] = None,
+        extract_xcom: bool = True
     ):
         self.validate_pod_generator_args(locals())
```
`validate_pod_generator_args` can be significantly simplified. The only reason it was so complex was because of the signature inspection it was doing for the ridiculous number of arguments being passed. Now the validation is quite simple, the client either passes `pod` or `pod_template_file`.

dimberman replied:
@davlum good point, fixed!

davlum reviewed (commented):
If possible I would further simplify this by removing all of the `reconcile_*` methods as they can have unexpected results for the client. The attributes Airflow wants to add at runtime (what's currently called the dynamic pod) do not actually constitute a full pod definition. They are limited and their interactions fairly well understood. The possibilities on configuration now shouldn't have to merge arbitrary pods, which can be troublesome.

The options are as I understand it are;
1. The client configures the pod via `airflow.cfg`, and doesn't pass `executor_config`. Just add the additional dynamic arguments to the pod.
2. The client configures a pod via `airflow.cfg` but does pass a pod via `executor_config`. Ignore the `airflow.cfg` pod, use the `executor_config` pod instead and add the dynamic arguments to that pod.

kaxil commented on UPDATING.md:
```diff
@@ -154,6 +153,480 @@ The Old and New provider configuration keys that have changed are as follows
+In Airflow 2.0, the KubernetesExecutor will require a base pod template written in yaml. This file can exist
+anywhere on the host machine and will be linked using the `pod_template_file` configuration in the airflow.cfg.
+
+The airflow.cfg will still accept values for the `worker_container_repository`, the `worker_container_tag`, and
```
```suggestion
The `airflow.cfg` will still accept values for the `worker_container_repository`, the `worker_container_tag`, and
```

kaxil commented on UPDATING.md:
```diff
@@ -154,6 +153,480 @@ The Old and New provider configuration keys that have changed are as follows
+The airflow.cfg will still accept values for the `worker_container_repository`, the `worker_container_tag`, and
+the default namespace.
+
+#### The executor_config Will Now Expect a `kubernetes.client.models.V1Pod` Class When Launching Tasks
```
```suggestion
#### The `executor_config` Will Now Expect a `kubernetes.client.models.V1Pod` Class When Launching Tasks
```

kaxil commented on UPDATING.md:
```diff
@@ -154,6 +153,480 @@ The Old and New provider configuration keys that have changed are as follows
+
+#### The executor_config Will Now Expect a `kubernetes.client.models.V1Pod` Class When Launching Tasks
+
+In airflow 1.10, users could modify task pods at runtime by passing a dictionary to the `executor_config` variable.
```
```suggestion
In Airflow 1.10.x, users could modify task pods at runtime by passing a dictionary to the `executor_config` variable.
```

kaxil commented on UPDATING.md:
```diff
@@ -154,6 +153,480 @@ The Old and New provider configuration keys that have changed are as follows
+
+### Changes to the KubernetesPodOperator
+
+Much like the KubernetesExecutor, the KubernetesPodOperator will no longer take Airflow custom classes and will
```
```suggestion
Much like the `KubernetesExecutor`, the `KubernetesPodOperator` will no longer take Airflow custom classes and will
```

dimberman committed:
fix documentation

dimberman committed:
simplify validate function

mik-laj commented on UPDATING.md:
```diff
@@ -154,6 +153,480 @@ The Old and New provider configuration keys that have changed are as follows
 
+### Changes to the KubernetesExecutor
+
+#### The KubernetesExecutor Will No Longer Read from the airflow.cfg for Base Pod Configurations
```
Can you add a full list of configuration options that have been deleted? This is important if we are going to work on `airflow upgrade-check`, because having the list we can detect incorrect configuration and notify the user at runtime.

dimberman replied:
@mik-laj added

mik-laj commented on UPDATING.md:
````diff
@@ -154,6 +153,480 @@ The Old and New provider configuration keys that have changed are as follows
+    }
+)
+```
+For Airflow 2.0, the traditional `executor_config` will continue operation with a deprecation warning,
````
Do we have this deprecation warning? I didn't see it in the code, but I might have missed it.

dimberman replied:
Hi @mik-laj, the deprecation warning was already merged so it didn't show up in the PR https://github.com/apache/airflow/pull/10393/files#diff-868ed785b2a336f20cb6a577dde2502aR198-R201

mik-laj commented on docs/howto/operator/kubernetes.rst:
```diff
@@ -53,13 +53,10 @@ Ultimately, it allows Airflow to act a job orchestrator - no matter the language
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-Kubernetes cluster resources such as ConfigMaps, Secrets, and Volumes can be used with a Pod to be launched.
+Kubernetes cluster resources such as Secrets can be used with a Pod to be launched.
 Utilize the Airflow Kubernetes model classes such as:
```
This sentence needs to be corrected. The plural is used when one class is described.

ashb replied:
Do we still support standard python dicts? (Line 60)

Also: why did this bit need to change -- should `Volume`  or `VolumeMount` not be used anymore?

dimberman replied:
Fixed. @ashb yes `Volume` and `VolumeMount` would now be `k8s.V1Volume` and `k8s.V1VolumeMount`, I go into more detail on this in the UPDATING.md

dimberman committed:
@mik-laj comments

dimberman committed:
spellcheck

dimberman committed:
spellcheck

ashb commented on airflow/kubernetes/pod_generator.py:
```diff
@@ -396,7 +240,7 @@ def from_legacy_obj(obj) -> Optional[k8s.V1Pod]:
                 )
         namespaced['resources'] = resources
-        return PodGenerator(**namespaced).gen_pod()
+        return PodGeneratorDeprecated(**namespaced).gen_pod()
```
Can/should we issue a FutureDeprecationWarning here, so that we can remove this code path in 2.1/2.2?

dimberman replied:
@ashb there is a future deprecation https://github.com/apache/airflow/pull/10393/files#diff-868ed785b2a336f20cb6a577dde2502aR198-R201

ashb replied:
Oh, you issue a warning a few lines up (from the previous PR. right?)

dimberman replied:
Yep :)

dimberman committed:
Update airflow/executors/kubernetes_executor.py

Co-authored-by: Kaxil Naik <kaxilnaik@gmail.com>
